adapters/matterfull/matterfull.go

```diff
@@ -0,0 +1,221 @@
+	var errors []error
+	resImps := make([]openrtb2.Imp, 0, len(imps))
+	res := make(map[openrtb_ext.ExtImpMatterfull][]openrtb2.Imp)
+
+	for imp := range iterutil.SlicePointerValues(imps) {
+		impExt, err := getImpressionExt(imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		//dispatchImpressions
+		//Group impressions by Matterfull-specific parameters `pid
+		if err := compatImpression(imp); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		res[impExt] = append(res[impExt], *imp)
+		resImps = append(resImps, *imp)
+	}
+	return res, resImps, errors
+}
+
+// Alter impression info to comply with Matterfull platform requirements
+func compatImpression(imp *openrtb2.Imp) error {
+	imp.Ext = nil //do not forward ext to Matterfull platform
+	if imp.Banner != nil {
+		return compatBannerImpression(imp)
+	}
+	return nil
+}
+
+func compatBannerImpression(imp *openrtb2.Imp) error {
+	// Create a copy of the banner, since imp is a shallow copy of the original.
+
+	bannerCopy := *imp.Banner
+	banner := &bannerCopy
+	//As banner.w/h are required fields for Matterfull platform - take the first format entry
+	if banner.W == nil || banner.H == nil {
+		if len(banner.Format) == 0 {
+			return &errortypes.BadInput{Message: "Expected at least one banner.format entry or explicit w/h"}
+		}
+		format := banner.Format[0]
+		banner.Format = banner.Format[1:]
+		banner.W = &format.W
+		banner.H = &format.H
+		imp.Banner = banner
+	}
+	return nil
+}
+
+func getImpressionExt(imp *openrtb2.Imp) (openrtb_ext.ExtImpMatterfull, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := jsonutil.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return openrtb_ext.ExtImpMatterfull{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	var MatterfullExt openrtb_ext.ExtImpMatterfull
+	if err := jsonutil.Unmarshal(bidderExt.Bidder, &MatterfullExt); err != nil {
+		return openrtb_ext.ExtImpMatterfull{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	return MatterfullExt, nil
+}
+
+func (adapter *adapter) buildAdapterRequest(prebidBidRequest *openrtb2.BidRequest, params *openrtb_ext.ExtImpMatterfull, imps []openrtb2.Imp) (*adapters.RequestData, error) {
+	newBidRequest := createBidRequest(prebidBidRequest, params, imps)
+	reqJSON, err := jsonutil.Marshal(newBidRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	url, err := adapter.buildEndpointURL(params)
+	if err != nil {
+		return nil, err
+	}
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     url,
+		Body:    reqJSON,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(imps)}, nil
+}
+
+func createBidRequest(prebidBidRequest *openrtb2.BidRequest, params *openrtb_ext.ExtImpMatterfull, imps []openrtb2.Imp) *openrtb2.BidRequest {
+	bidRequest := *prebidBidRequest
```

Comment:
The bid request being passed to your adapter is already a shallow copy of the original bid request unique to your adapter. So I don't think you need to make another shallow copy of the request. Since the `Site` and `App` edits are consistent, you could just execute this once at the beginning rather that execute this for each impression group. Its just the imps assignment that is unique to each impression group. In any case, you do need to make the `Site` and `App` copies that you do below, as those are shared across all adapters.